src/services/htmlCompletion.ts

```diff
@@ -9,407 +9,435 @@ import { HTMLDocument, Node } from '../parser/htmlParser';
+			let range: Range;
+			let addQuotes: boolean;
+			if (offset > valueStart && offset <= valueEnd && text[valueStart] === '"') {
+				// inside attribute
+				if (valueEnd > offset && text[valueEnd - 1] === '"') {
+					valueEnd--;
+				}
+				let wsBefore = getWordStart(text, offset, valueStart + 1);
+				let wsAfter = getWordEnd(text, offset, valueEnd);
+				range = getReplaceRange(wsBefore, wsAfter);
+				addQuotes = false;
+			} else {
+				range = getReplaceRange(valueStart, valueEnd);
+				addQuotes = true;
 			}
-			let wsBefore = getWordStart(text, offset, valueStart + 1);
-			let wsAfter = getWordEnd(text, offset, valueEnd);
-			range = getReplaceRange(wsBefore, wsAfter);
-			addQuotes = false;
-		} else {
-			range = getReplaceRange(valueStart, valueEnd);
-			addQuotes = true;
-		}
-		let tag = currentTag.toLowerCase();
-		let attribute = currentAttributeName.toLowerCase();
-		tagProviders.forEach(provider => {
-			provider.collectValues(tag, attribute, value => {
-				let insertText = addQuotes ? '"' + value + '"' : value;
-				result.items.push({
-					label: value,
-					filterText: insertText,
-					kind: CompletionItemKind.Unit,
-					textEdit: TextEdit.replace(range, insertText),
-					insertTextFormat: InsertTextFormat.PlainText
+
+			let tag = currentTag.toLowerCase();
+			let attribute = currentAttributeName.toLowerCase();
+			let value = extractAttributeValue(scanner.getTokenText());
+
+			completionParticipants.forEach(participant => {
+				if (participant.onHtmlAttributeValue) {
+					participant.onHtmlAttributeValue({ tag, attribute, value, range });
+				}
+      });
+
+			tagProviders.forEach(provider => {
+				provider.collectValues(tag, attribute, value => {
+					let insertText = addQuotes ? '"' + value + '"' : value;
+					result.items.push({
+						label: value,
+						filterText: insertText,
+						kind: CompletionItemKind.Unit,
+						textEdit: TextEdit.replace(range, insertText),
+						insertTextFormat: InsertTextFormat.PlainText
+					});
 				});
 			});
-		});
-		collectCharacterEntityProposals();
-		return result;
-	}
+			collectCharacterEntityProposals();
+			return result;
+		}
 
-	function scanNextForEndPos(nextToken: TokenType): number {
-		if (offset === scanner.getTokenEnd()) {
-			token = scanner.scan();
-			if (token === nextToken && scanner.getTokenOffset() === offset) {
-				return scanner.getTokenEnd();
+		function scanNextForEndPos(nextToken: TokenType): number {
+			if (offset === scanner.getTokenEnd()) {
+				token = scanner.scan();
+				if (token === nextToken && scanner.getTokenOffset() === offset) {
+					return scanner.getTokenEnd();
+				}
 			}
+			return offset;
 		}
-		return offset;
-	}
 
-	function collectInsideContent(): CompletionList {
-		return collectCharacterEntityProposals();
-	}
+		function collectInsideContent(): CompletionList {
+			completionParticipants.forEach(participant => {
```

Comment:
I'd prefer `for (let participant of completionParticipants) {` (easier to debug and avoids the need to create a function)